### PR TITLE
even lighter version of #17938: fix most issues with UnusedImport, XDeclaredButNotUsed, etc; fix #17511, #17510, #14246 (without realModule)

### LINKED ISSUE #14246: Incorrect `UnusedImport`  warning with re-exported symbols or with  `tyAlias` (`type Foo* = Bar`)
### Example
```nim
import strutils as su
export su
```

### Current Output
```
Warning: imported and not used: 'strutils' [UnusedImport]
```
Ps:
```nim
import strutils
export strutils
```
Don't have this warnning.

```
$ nim -v
Nim Compiler Version 1.2.0
```

### LINKED ISSUE #17510: Incorrect `UnusedImport` warnings regression from 1.4.2 to 1.5.1
When one module imports and exports a module, another module depending on the imported/exported module claims in 1.5.1 that the intermediate `import`/`export`ing module is unused.

### Example
`test.nim`:
```nim
import sequtils
export sequtils
```
`test2.nim`:
```nim
import "."/test
echo repeat("b", 3)
```

### Current Output
With head:
```
$ ~/Nim/bin/nim c -r test2.nim
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
.......
test2.nim(1, 11) Warning: imported and not used: 'test' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: test2.nim
Hint:  [Link]
Hint: 25592 lines; 0.544s; 31.934MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ ~/Nim/bin/nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 8573160a44439ffe3fbbdd354188db9c976a2081
active boot switches: -d:release
```

Nim 1.4.2 does not appear to emit this incorrect warning:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
.......CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim

Hint:  [Link]
Hint: 25242 lines; 0.534s; 25.52MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-12-02
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
$ 
```
Nor does 1.0.4.

### Expected Output
No incorrect `UnusedImport` warning

### LINKED ISSUE #17511: Incorrect `UnusedImport` warning in presence of `import foo as bar`
When module `bar` imports and exports module `foo` (here, `sequtils`) using `import ... as ...` and module `baz` imports `bar` to get access to `foo`, either one or two `UnusedImport` warning is emitted, depending on Nim version.

### Example
`test.nim`:
```nim
import sequtils as su2
export su2
```

`test2.nim`:
```nim
import "."/test
echo repeat("b", 3)
```

### Current Output
```
$ ~/Nim/bin/nim c -r test2.nim
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
.......
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
test2.nim(1, 11) Warning: imported and not used: 'test' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: test2.nim
Hint:  [Link]
Hint: 25592 lines; 0.530s; 31.934MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
user@tertiva:~$ ~/Nim/bin/nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 8573160a44439ffe3fbbdd354188db9c976a2081
active boot switches: -d:release
```

Older versions also display variations of this behavior.
1.4.2:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
.......
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: 25242 lines; 0.542s; 25.52MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-12-02
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```

1.2.6:
```
$ Nim1.2.6/bin/nim c -r test2.nim
Hint: used config file 'Nim1.2.6/config/nim.cfg' [Conf]
Hint: used config file 'Nim1.2.6/config/config.nims' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: test2 [Processing]
Hint: test [Processing]
Hint: sequtils [Processing]
Hint: macros [Processing]
/home/user/test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: 25264 LOC; 0.532 sec; 30.262MiB peakmem; Debug build; proj: test2.nim; out: test2 [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ Nim1.2.6/bin/nim --version
Nim Compiler Version 1.2.6 [Linux: amd64]
Compiled at 2021-03-24
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: f05ba3c667157f46f141ff700121c7aff86cf07a
active boot switches: -d:release
```

1.0.4:
```
$ nim c -r test2.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: test2 [Processing]
Hint: test [Processing]
Hint: sequtils [Processing]
Hint: macros [Processing]
test.nim(1, 17) Warning: imported and not used: 'sequtils' [UnusedImport]
CC: stdlib_assertions.nim
CC: stdlib_dollars.nim
CC: stdlib_io.nim
CC: stdlib_system.nim
CC: stdlib_sequtils.nim
CC: test2.nim
Hint:  [Link]
Hint: operation successful (25045 lines compiled; 0.523 sec total; 29.773MiB peakmem; Debug Build) [SuccessX]
Hint: test2  [Exec]
@["b", "b", "b"]
$ nim --version
Nim Compiler Version 1.0.4 [Linux: amd64]
Compiled at 2020-02-04
Copyright (c) 2006-2019 by Andreas Rumpf

active boot switches: -d:release
```

### Expected Output
No incorrect `UnusedImport` warning(s).

### Additional Information
https://github.com/nim-lang/Nim/issues/17510 is related, but only appears after Nim 1.4.2 and doesn't need `import foo as bar` / `export bar`, just `import foo` / `export foo`.

### PR DESCRIPTION
**Pull request by timotheecour**
* 3rd attempt; lighter version of https://github.com/nim-lang/Nim/pull/17978 and https://github.com/nim-lang/Nim/pull/17938

the main difference with https://github.com/nim-lang/Nim/pull/17978 is it doesn't add a `realModule` field

fix #17511
fix #17510
fix #14246
closes https://github.com/nim-lang/Nim/pull/17978

## future work
- [ ] after this, consider setting `switch("warningAsError", "UnusedImport")` in `nimStrictMode`
- [ ] revisit https://github.com/nim-lang/Nim/pull/17938

### CODE REVIEW

File: compiler/suggest.nim [hunk 1]
```diff
@@ -572,7 +572,8 @@ proc markOwnerModuleAsUsed(c: PContext; s: PSym) =
     var i = 0
     while i <= high(c.unusedImports):
       let candidate = c.unusedImports[i][0]
-      if candidate == module or c.exportIndirections.contains((candidate.id, s.id)):
+      let candidate2 = candidate # PRTEMP
```
Comment (timotheecour) [inline compiler/suggest.nim]: i know; I will simplify later

File: lib/system/schubfach.nim [hunk 2]
```diff
@@ -95,6 +95,36 @@ proc floorLog2Pow10(e: int32): int32 {.inline.} =
   sf_Assert(e <= 1233)
   return floorDivPow2(e * 1741647, 19)
 
+const
```
Comment (Araq) [inline lib/system/schubfach.nim:98]: Why was this required? Workaround a pre-existing IC bug?

Comment (timotheecour) [reply lib/system/schubfach.nim:98]: yes, this is explained here https://github.com/nim-lang/Nim/pull/17978#issuecomment-868892830

File: compiler/suggest.nim (hunk 1 above)
Comment (timotheecour) [reply compiler/suggest.nim]: cleaned up

File: compiler/lineinfos.nim [hunk 3]
```diff
@@ -67,6 +67,7 @@ type
     warnResultUsed = "ResultUsed",
     warnCannotOpen = "CannotOpen",
     warnFileChanged = "FileChanged",
+    warnDuplicateModuleImport = "DuplicateModuleImport",
```
Comment (Clyybber) [inline compiler/lineinfos.nim:70]: If at all this should be a hint, not a warning.

**Comment by timotheecour** [comment]
> If at all this should be a hint, not a warning.

I'm ok changing to a hint, given that we have hintAsError

This PR generates a lot of valid warnings when compiling nim, eg:
* ` Warning: imported and not used: 'platform' [UnusedImport]`)
* docgen.nim(19, 14) Warning: redefinition of 'trees'; previous declaration here [...]

but also a few bad ones in cases like this:

```nim
import strutils except Letters
from os import nil
```
need to followup with a PR to handle those cases => https://github.com/nim-lang/Nim/pull/18366

File: compiler/lineinfos.nim (hunk 3 above)
Comment (timotheecour) [reply compiler/lineinfos.nim:70]: => doing this in https://github.com/nim-lang/Nim/pull/18366

**Pull request merged**